### PR TITLE
Ecdh curves support for non ISTIO_MUTUAL traffic

### PR DESCRIPTION
**Pull request by kfaseela**
API PR: https://github.com/istio/api/pull/2543
Fixes: https://github.com/istio/istio/issues/41645

### CODE REVIEW

**Review by GregHanson — CHANGES_REQUESTED** [review]
[this](https://github.com/istio/istio/issues/41645) seems like a new feature/API, can you add release note linking the issue or istio/api PR?

File: pkg/config/security/security.go
Comment (bleggett) [inline pkg/config/security/security.go:232]: This is a little confusing to me - if no curves are specified, that's a valid curve?

What's the expected/desired outcome in this case? All curves allowed? No curves allowed?

If this does have special meaning, an explicit test for no curves specified would be good.

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (bleggett) [inline pilot/pkg/networking/core/v1alpha3/cluster_builder.go:1096]:
```suggestion
		ApplyECDHCurves(tlsContext, opts.mesh)
```

nit - should be ECDH everywhere

File: pkg/config/security/security.go
Comment (bleggett) [inline pkg/config/security/security.go:211]:
```suggestion
var ValidECDHCurves = sets.New(
```

nit: should be ECDH in all naming

File: pkg/config/security/security.go
Comment (bleggett) [inline pkg/config/security/security.go:240]:
```suggestion
func FilterECDHCurves(suites []string) []string {
```

nit: ditto

File: pkg/config/security/security.go
Comment (kfaseela) [reply pkg/config/security/security.go]: MeshConfig ECDH curves are validated only if atleast one curve is specified. If no curves are specified it follows the current default behavior for backward compatibility - which means we let envoy decide the default values for us, as specified in : https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto

Comment (bleggett) [reply pkg/config/security/security.go]: Makes sense - a comment indicating that for this function would be good, as it's not entirely obvious that is what would happen from this code.

**Comment by GregHanson** [comment]
> I added release notes in the API PR. Isn't that enough? https://github.com/istio/api/pull/2543/files

good question, @ericvn are istio/api release notes picked up by the release notes script too?

File: pkg/config/security/security.go
Comment (kfaseela) [reply pkg/config/security/security.go:211]: From Istio we will not set any default value because of this, and would let envoy choose the default. However if the control plane wants to choose a specific curve against letting envoy choose the default, we allow this config option. Did I miss any supported curve here in the validation, which will be otherwise allowed by envoy?

Comment (kyessenov) [reply pkg/config/security/security.go:211]: Well, you are looking at a particular commit of boringssl. Curves can be added and removed at any time (a vulnerability found, US government adds it to FIPS approved list, etc). It depends on how envoy was built (which version of boringssl/openssl/fips was used). I am worried this list can become actively harmful later, because users will not be able to use the right curve.

I think the list of curves should not be built into code. It needs to be vendor-specific (US companies might want FIPS list, while the rest of the world is fine with Wireguard's CHACHA) as well as Envoy-build specific (so that we can transition between Envoy versions). 

Sending a wrong curve I believe will lead Envoy deny the entire SDS resource.

Comment (kfaseela) [reply pkg/config/security/security.go:211]: @kyessenov : Would it be better if I remove the validation from Istio, and leave the validation responsibility to envoy?

**Comment by ericvn** [comment]
> > I added release notes in the API PR. Isn't that enough? https://github.com/istio/api/pull/2543/files
> 
> good question, @ericvn are istio/api release notes picked up by the release notes script too?

That was a good question so I went to look at what the tool does. It processes a specified repo at a time, so you could specify istio or api or potentially any other repo with release notes (istio and api are the only two with a releasenotes directory today). I then went to look at what the RM instructions might be relative to creating release notes. Do we instruct them to run it against both the istio and api repos? The wikis don't seem to mention one or both.

@GregHanson, as a prior RM when the tooling was available (at least I believe you fit that), was the tooling run against both?

Also as a side note, we do run the release notes required tooling against both repos, however in the api repo it is not a required test so PRs will merge even if the test fails. Most people might not notice the `required` label.

So what's the best way forward, and we should update the RM wikis and/or READMEs for this? If we really have `releasenotes` directories in both repos, the RMs should likely be running the tooling against both repos. I haven't looked at the PRs to actually see if the api repo is used or not, but I suspect there are PRs that only happen in the api repo, so we need to scan those as well. For those api PRs that the creator knows needs an istio change, it makes sense to include the release note in one or the other (including in both will just mean the RM can remove the duplicate when combining). It may make sense to only include in istio as the api PR will merge without it (maybe add a remark in the api PR that the RN will be in istio).

File: pkg/config/security/security.go
Comment (kyessenov) [reply pkg/config/security/security.go:211]: Yes, I'm fine with no validation. Basically, half-working validation is more harmful than no validation.

Comment (kyessenov) [reply pkg/config/security/security.go:211]: +1 for warning, but noone should by typing manually crypto-related settings :)
They should be in gitops.

**Comment by kfaseela** [comment]
> > > I added release notes in the API PR. Isn't that enough? https://github.com/istio/api/pull/2543/files
> > 
> > 
> > good question, @ericvn are istio/api release notes picked up by the release notes script too?
> 
> That was a good question so I went to look at what the tool does. It processes a specified repo at a time, so you could specify istio or api or potentially any other repo with release notes (istio and api are the only two with a releasenotes directory today). I then went to look at what the RM instructions might be relative to creating release notes. Do we instruct them to run it against both the istio and api repos? The wikis don't seem to mention one or both.
> 
> @GregHanson, as a prior RM when the tooling was available (at least I believe you fit that), was the tooling run against both?
> 
> Also as a side note, we do run the release notes required test in the CI for both repos, however in the api repo it is not a required test so PRs will merge even if the test fails. Most people might not notice the `required` label.
> 
> So what's the best way forward, and we should update the RM wikis and/or READMEs for this? If we really have `releasenotes` directories in both repos, the RMs should likely be running the tooling against both repos. I haven't looked at the PRs to actually see if the api repo is used or not, but I suspect there are PRs that only happen in the api repo, so we need to scan those as well. For those api PRs that the creator knows needs an istio change, it makes sense to include the release note in one or the other (including in both will just mean the RM can remove the duplicate when combining). It may make sense to only include in istio as the api PR will merge without it (maybe add a remark in the api PR that the RN will be in istio).

@GregHanson Meantime I copied the release notes to this PR :D

File: pkg/config/security/security.go
Comment (kfaseela) [reply pkg/config/security/security.go]: @bleggett I added this to the description of the field in the API PR, so that people who knows the config knows how to use it and what the defaults are

File: go.mod
Comment (kfaseela) [reply go.mod]: yes

File: pkg/config/security/security.go
Comment (kfaseela) [reply pkg/config/security/security.go:211]: Changed the errors to warnings

File: pkg/config/security/security.go
Comment (bleggett) [reply pkg/config/security/security.go]: That works, thx! As long as it's breadcrumb'd.

File: pkg/config/validation/validation.go
Comment (bleggett) [inline pkg/config/validation/validation.go:1755]:
```suggestion
		v = appendWarningf(v, "detected unrecognized ECDH curves: %v", sets.SortedList(invalidEcdhCurves))
```

This wording might be less confusing, since we don't actually know if they're invalid, only Envoy knows that, they just don't match our hardcoded validation list.

**Comment by dhawton** [comment]
> > > I added release notes in the API PR. Isn't that enough? https://github.com/istio/api/pull/2543/files
> > 
> > 
> > good question, @ericvn are istio/api release notes picked up by the release notes script too?
> 
> That was a good question so I went to look at what the tool does. It processes a specified repo at a time, so you could specify istio or api or potentially any other repo with release notes (istio and api are the only two with a releasenotes directory today). I then went to look at what the RM instructions might be relative to creating release notes. Do we instruct them to run it against both the istio and api repos? The wikis don't seem to mention one or both.
> 
> @GregHanson, as a prior RM when the tooling was available (at least I believe you fit that), was the tooling run against both?
> 
> Also as a side note, we do run the release notes required test in the CI for both repos, however in the api repo it is not a required test so PRs will merge even if the test fails. Most people might not notice the `required` label.
> 
> So what's the best way forward, and we should update the RM wikis and/or READMEs for this? If we really have `releasenotes` directories in both repos, the RMs should likely be running the tooling against both repos. I haven't looked at the PRs to actually see if the api repo is used or not, but I suspect there are PRs that only happen in the api repo, so we need to scan those as well. For those api PRs that the creator knows needs an istio change, it makes sense to include the release note in one or the other (including in both will just mean the RM can remove the duplicate when combining). It may make sense to only include in istio as the api PR will merge without it (maybe add a remark in the api PR that the RN will be in istio).

I can say that I have only run it against the istio repo and not both as there didn't seem to be any instruction to run against both repos.

**Review by dhawton — APPROVED** [review]
Release notes look good.

File: pkg/config/validation/validation.go
Comment (kfaseela) [reply pkg/config/validation/validation.go]: done

Comment (bleggett) [reply pkg/config/validation/validation.go]: Thanks!

**Review by bleggett — APPROVED** [review]

**Review by GregHanson — APPROVED** [review]
Thanks @kfaseela !

File: pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go:2677]: Can you please also add a test case for MUTUAL?

File: pilot/pkg/networking/core/v1alpha3/gateway_test.go
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/gateway_test.go:683]: Same here. Can you add a test case for MUTUAL?

File: pilot/pkg/networking/core/v1alpha3/listener.go
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/listener.go:199]: I think this has to be set only if len(EcdhCurves) > 0. Otheriwse it will overwrite the Envoys default Ecdh curves

**Comment by jacob-delgado** [comment]
LGTM, but I'll wait to approve once istio/api PR merges.

File: pilot/pkg/networking/core/v1alpha3/gateway_test.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/gateway_test.go:683]: Added additional test cases

File: pilot/pkg/networking/core/v1alpha3/listener.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/listener.go:199]: I tested this out and this seems to be working as expected.

File: pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go:2677]: Added additional test cases

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go:2680]: Sorry I meant to ask ISTIO_MUTUAL as well. If Mesh config has ecdh curves configured, for Istio Mutual we should not add. Can you please add that test also? Here and below?

**Review by ramaraochavali — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster_builder.go:1200]: We should validate the values to prevent sending invalid curves to envoy

File: pilot/pkg/networking/core/v1alpha3/listener.go
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/listener.go:210]: Doubt why MeshExternal_TLS should be applied to listener

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:1200]: @hzxuzhonghu since ecdh curves is in Meshconfig, the validation is added while configuring MeshConfig.

**Review by ericvn — APPROVED** [review]
Tests and release notes look good.

File: pilot/pkg/networking/core/v1alpha3/listener.go
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener.go:210]: I think it is needed for gateways receiving external traffic. Are you thinking we should only this for gateways and not for sidecar inbound?

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener.go:210]: I think his point still valid - Should we only set when we are building listener for gateway and not sidecar?

Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/listener.go:210]: @ramaraochavali i think it is valid for external traffic still. We do have usecases where we donot use gateway, and uses sidercar TLS termination for external traffic using sidecar API

File: pkg/config/security/security.go
Comment (irisdingbj) [inline pkg/config/security/security.go:210]: Do we only suppot the one form boringssl? any conflict with openssl?

File: pilot/pkg/networking/core/v1alpha3/listener.go
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/listener.go:210]: But in this case it will also take effect on sidecar->sidecar communication

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener.go:210]: yes. That is the side affect of this. We apply the same settings for cluster outbound. So while the setting says MeshExternal, in a way we are applying it for mesh calls also. That is the confusion I was talking about [here](https://github.com/istio/istio/pull/43207#discussion_r1112156882).

I agree though that it is very confusing implementation.

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener.go:210]: I just added DNM label so that we can sort this out.

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener.go:210]: Added this https://github.com/istio/api/pull/2543/files#r1112668620 in API.

**Review by ramaraochavali — CHANGES_REQUESTED** [review]
Let us sort out the impact on sidear -> sidecar with this setting

**Comment by ramaraochavali** [comment]
@howardjohn @hzxuzhonghu PTAL - if you are OK with these changes, I will remove DNM

**Comment by howardjohn** [comment]
im ok

**Review by howardjohn — APPROVED** [review]

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
I donot feel it quite do as the api means.  Take `cluster` as an example, it applies mesh external tls settings to cluster. 
But the cluster can not be distinguished from internal to external well.

**Comment by kfaseela** [comment]
> I donot feel it quite do as the api means. Take `cluster` as an example, it applies mesh external tls settings to cluster. But the cluster can not be distinguished from internal to external well.

@hzxuzhonghu : I have added API docs update as we discussed in the discussion above, let me know if that solves your concern: https://github.com/istio/api/pull/2693/files

**Review by ramaraochavali — APPROVED** [review]
I have noted my preference [here](https://github.com/istio/api/pull/2693#pullrequestreview-1314464164).

**Comment by kfaseela** [comment]
@hzxuzhonghu : I have not been able to change the API to have a single mesh_mTLS config, as @howardjohn also indicates that it is currently not applied in all contexts. https://github.com/istio/api/pull/2693/files#r1118972237 .  If you want to still retain the hold on this PR, please suggest a way forward.

File: pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder_test.go:2697]: Done

File: pilot/pkg/networking/core/v1alpha3/listener.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/listener.go:207]: As per the discussions, API has been renamed to "tlsDefaults", instead of "external_tls". The implementation PR has been updated to use the new APIs

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster_builder.go:1175]: Not right here, you should exclude `networking.ClientTLSSettings_ISTIO_MUTUAL`

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster_builder.go:1187]: You can make TlsDefaults as a prama.

And from the discussion I think tls version should also be applied

File: pilot/pkg/networking/core/v1alpha3/gateway.go
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:631]: nit: pass tlsdefaults?

File: pilot/pkg/networking/core/v1alpha3/listener.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/listener.go:204]: @ramaraochavali : as per our discussion, in this PR I was supposed to focus only on ecdh curves. and you were planning to put a separate PR for TLS Version?

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/listener.go:204]: @hzxuzhonghu I will fix the minProtocolVersion part in a separate PR

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (kfaseela) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:1187]: @hzxuzhonghu : There are several unit tests with meshconfig not initialized, so if I change the param to tlsDefaults, i will end up making the nil checks from the caller, which is currently now in two different places. Does it not look neater the current way?

File: pilot/pkg/networking/core/v1alpha3/cluster_builder.go
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster_builder.go:1175]: IC

**Review by hzxuzhonghu — COMMENTED** [review]
LGTM

Can do cleanups in a following pr

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by kfaseela** [comment]
> LGTM
> 
> Can do cleanups in a following pr

thanks @hzxuzhonghu :)

**Pull request merged**
